### DISCUSSION TITLE
Any tips how to get C# .NET support for wasm working in this runtime?

### DISCUSSION THREAD

**Discussion by schaveyt** [WasmEdge/WasmEdge#959] [Q&A] [answered]
With the advent of .NET 5, Microsoft was able compile the .net runtime into a small WASM library with which standard .NET compiled IL libraries would run in. Then .NET 6, they made a lot of progress with getting Ahead-Of-Time (AOT) compiling for WASM. 

I would be interested in understanding how I could use those recent developments to get something running in the wasmedge runtime.

**Comment by hydai** [comment]
Hi @schaveyt 

If you are talking about running the application on wasm runtime:

The .NET WebAssembly build tools is based on Emscripten. And it also provides an `dotnet.wasm` which will take the the .NET internal runtime. And it's possible to compile the .Net program and link the `dotnet.wasm` statically to generate a standalone wasm file.

Then, you can use WasmEdge to execute this standalone wasm file.


Or you want to embed the wasm execution engine into the C# application?

**Reply by schaveyt** [reply]
Apparently, @SteveSandersonMS, the legend, has been prototyping the very thing I'm looking for. See https://youtu.be/8gwSU3oaMV8?t=01h01m51s

It sounds like it is in it's infancy but the team clearly see the significance/potential of the wasm runtime approach

**Reply by hydai** [reply]
Sorry for the late response.

In the prototyping you mentioned, the .Net runtime and application can be compiled into a wasm file (called `console.wasm` in the demo) and a loadable dll file (I think this is the runtime itself). And then use a wasm runtime to run it directly.

It's possible to run it with WasmEdge just like the `iwasm` and `wasmer` in the video.

**Comment by SteveSandersonMS** [comment, marked as answer]
The prototype-level support was merged into the `dotnet/runtime` repo. You can find a README.md inside https://github.com/dotnet/runtime/tree/main/src/mono/wasi giving steps on how to build and run.

**Reply by SteveSandersonMS** [reply]
Note that, for reasons I don't know, it runs *much* slower on WasmEdge than on Wasmtime, Wasmer, or WAMR. It might just be the initial startup that's slower - I'm not sure.

**Reply by hydai** [reply]
Thanks for the information.
I am going to try it to figure out the reason.
